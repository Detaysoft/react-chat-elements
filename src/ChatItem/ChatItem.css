.rce-container-citem {
    flex-direction: column;
    display: block;
    overflow: hidden;
    min-width: 240px;
}

.rce-citem {
    position: relative;
    background: white;
    display: flex;
    flex-direction: row;
    height: 72px;
    cursor: pointer;
    user-select: none;
    max-width: 100%;
    overflow: hidden;
    min-width: 240px;
}

.rce-citem:hover {
    background: #f9f9f9;
}

.rce-citem-avatar {
    position: relative;
    padding: 0 15px 0 13px;
    justify-content: center;
    display: flex;
    align-items: center;
}

.rce-citem-status {
    width: 20px;
    height: 20px;
    bottom: 10px;
    right: 10px;
    position: absolute;
    border-radius: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    background: #ccc;
}

.rce-citem-avatar img {
    width: 50px;
    height: 50px;
    border: none !important;
    background: #ccc;
    border-radius: 100%;
    overflow: hidden;
    font-size: 10px;
    text-align: center;
    line-height: 50px;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.rce-citem-body {
    display: flex;
    flex: 1;
    flex-direction: column;
    display: flex;
    justify-content: center;
    padding-right: 15px;
    border-bottom: 1px solid rgba(0,0,0,.05);
    overflow: hidden;
}

.rce-citem-body--top {
    display: flex;
}

.rce-citem-body--bottom {
    margin-top: 4px;
    display: flex;
}

.rce-citem-body--bottom-title, .rce-citem-body--top-title {
    flex: 1;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
}

.rce-citem-body--top-title {
    font-size: 16px;
}

.rce-citem-body--bottom-title {
    color: #555;
    font-size: 15px;
}

.rce-citem-body--top-time {
    font-size: 12px;
    color: rgba(0,0,0,0.4)
}

.rce-citem-body--bottom-status {
    margin-left: 3px;
}

.rce-citem-body--bottom-status span {
    width: 18px;
    height: 18px;
    font-size: 12px;
    color: white;
    font-weight: bold;
    text-align: center;
    align-items: center;
    justify-content: center;
    display: flex;
    border-radius: 100%;
    background: red;
}
