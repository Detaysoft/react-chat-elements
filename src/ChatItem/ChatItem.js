import React, { Component } from 'react';
import './ChatItem.css';

import Avatar from '../Avatar/Avatar';

const moment = require('moment')

export class ChatItem extends Component {

	render() {
		return (
			<div
				className="rce-container-citem"
				onClick={this.props.onClick}>
				<div className="rce-citem">
					<div className="rce-citem-avatar">
						<Avatar src={this.props.avatar} alt={this.props.alt} size="large" type="circle"/>
					</div>

					<div className="rce-citem-body">
						<div className="rce-citem-body--top">
							<div className="rce-citem-body--top-title">
								{this.props.title}
							</div>
							<div className="rce-citem-body--top-time">
								{moment(this.props.date).fromNow()}
							</div>
						</div>

						<div className="rce-citem-body--bottom">
							<div className="rce-citem-body--bottom-title">
								{this.props.subtitle}
							</div>
							<div className="rce-citem-body--bottom-status">
								{
									this.props.unread > 0 &&
									<span>{this.props.unread}</span>
								}
							</div>
						</div>
					</div>
				</div>
			</div>
		);
	}
}

ChatItem.defaultProps = {
	id: '',
	onClick: null,
	avatar: '',
	alt: '',
	title: '',
	subtitle: '',
	date: new Date(),
	unread: 0,
}

export default ChatItem;