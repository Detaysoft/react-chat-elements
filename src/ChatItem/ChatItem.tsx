import React, { Key, useEffect, useState } from 'react'
import './ChatItem.css'

import Avatar from '../Avatar/Avatar'

import { format } from 'timeago.js'

import classNames from 'classnames'

import { MdVideoCall, MdVolumeOff, MdVolumeUp } from 'react-icons/md'
import { IChatItemProps } from '../type'

const ChatItem: React.FC<IChatItemProps> = ({
  avatarFlexible = false,
  date = new Date(),
  unread = 0,
  statusColorType = 'badge',
  lazyLoadingImage = undefined,
  onAvatarError = () => void 0,
  ...props
}) => {
  const [onHoverTool, setOnHoverTool] = useState(false)
  const [onDrag, setOnDrag] = useState(false)

  useEffect(() => {
    props.setDragStates?.(setOnDrag)
  }, [])

  const handleOnMouseEnter = () => {
    setOnHoverTool(true)
  }

  const handleOnMouseLeave = () => {
    setOnHoverTool(false)
  }

  const handleOnClick = (e: React.MouseEvent) => {
    e.preventDefault()

    if (onHoverTool === true) return

    props.onClick?.(e)
  }

  const onDragOver = (e: React.MouseEvent) => {
    e.preventDefault()
    if (props.onDragOver instanceof Function) props.onDragOver(e, props.id)
  }

  const onDragEnter = (e: React.MouseEvent) => {
    e.preventDefault()
    if (props.onDragEnter instanceof Function) props.onDragEnter(e, props.id)
    if (!onDrag) setOnDrag(true)
  }

  const onDragLeave = (e: React.MouseEvent) => {
    e.preventDefault()
    if (props.onDragLeave instanceof Function) props.onDragLeave(e, props.id)
    if (onDrag) setOnDrag(false)
  }

  const onDrop = (e: React.MouseEvent) => {
    e.preventDefault()
    if (props.onDrop instanceof Function) props.onDrop(e, props.id)
    if (onDrag) setOnDrag(false)
  }

  return (
    <div
      key={props.id as Key}
      className={classNames('rce-container-citem', props.className)}
      onClick={handleOnClick}
      onContextMenu={props.onContextMenu}
    >
      <div className='rce-citem' onDragOver={onDragOver} onDragEnter={onDragEnter} onDragLeave={onDragLeave} onDrop={onDrop}>
        {!!props.onDragComponent && onDrag && props.onDragComponent(props.id)}
        {((onDrag && !props.onDragComponent) || !onDrag) && [
          <div
            key={'avatar'}
            className={classNames('rce-citem-avatar', { 'rce-citem-status-encircle': statusColorType === 'encircle' })}
          >
            <Avatar
              src={props.avatar}
              alt={props.alt}
              className={statusColorType === 'encircle' ? 'rce-citem-avatar-encircle-status' : ''}
              size='large'
              letterItem={props.letterItem}
              sideElement={
                props.statusColor ? (
                  <span
                    className='rce-citem-status'
                    style={
                      statusColorType === 'encircle'
                        ? {
                            border: `solid 2px ${props.statusColor}`,
                          }
                        : {
                            backgroundColor: props.statusColor,
                          }
                    }
                  >
                    {props.statusText}
                  </span>
                ) : (
                  <></>
                )
              }
              onError={onAvatarError}
              lazyLoadingImage={lazyLoadingImage}
              type={classNames('circle', { 'flexible': avatarFlexible })}
            />
          </div>,
          <div key={'rce-citem-body'} className='rce-citem-body'>
            <div className='rce-citem-body--top'>
              <div className='rce-citem-body--top-title'>{props.title}</div>
              <div className='rce-citem-body--top-time'>{date && (props.dateString || format(date))}</div>
            </div>

            <div className='rce-citem-body--bottom'>
              <div className='rce-citem-body--bottom-title'>{props.subtitle}</div>
              <div className='rce-citem-body--bottom-tools' onMouseEnter={handleOnMouseEnter} onMouseLeave={handleOnMouseLeave}>
                {props.showMute && (
                  <div className='rce-citem-body--bottom-tools-item' onClick={props.onClickMute}>
                    {props.muted === true && <MdVolumeOff />}
                    {props.muted === false && <MdVolumeUp />}
                  </div>
                )}
                {props.showVideoCall && (
                  <div className='rce-citem-body--bottom-tools-item' onClick={props.onClickVideoCall}>
                    <MdVideoCall />
                  </div>
                )}
              </div>
              <div className='rce-citem-body--bottom-tools-item-hidden-hover'>
                {props.showMute && props.muted && (
                  <div className='rce-citem-body--bottom-tools-item'>
                    <MdVolumeOff />
                  </div>
                )}
              </div>
              <div className='rce-citem-body--bottom-status'>{unread && unread > 0 ? <span>{unread}</span> : null}</div>
              {props.customStatusComponents !== undefined ? props.customStatusComponents.map(Item => <Item />) : null}
            </div>
          </div>,
        ]}
      </div>
    </div>
  )
}

export default ChatItem
