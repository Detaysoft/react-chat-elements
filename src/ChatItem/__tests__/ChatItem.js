import React, { Component } from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import ChatItem from '../ChatItem'

describe('ChatItem component', () => {
  it('should render without issues', () => {
    const component = shallow(<ChatItem avatar='' id='' />)

    expect(component.length).toBe(1)
    // TODO: Fix this test
    // expect(toJson(component)).toMatchSnapshot()
  })
})
