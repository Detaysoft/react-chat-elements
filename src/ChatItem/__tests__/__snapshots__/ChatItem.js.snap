// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatItem component should render without issues 1`] = `
<div
  className="rce-container-citem"
  onClick={null}
>
  <div
    className="rce-citem"
  >
    <div
      className="rce-citem-avatar"
    >
      <Avatar
        alt=""
        size="large"
        src=""
        type="circle"
      />
    </div>
    <div
      className="rce-citem-body"
    >
      <div
        className="rce-citem-body--top"
      >
        <div
          className="rce-citem-body--top-title"
        />
        <div
          className="rce-citem-body--top-time"
        >
          a few seconds ago
        </div>
      </div>
      <div
        className="rce-citem-body--bottom"
      >
        <div
          className="rce-citem-body--bottom-title"
        />
        <div
          className="rce-citem-body--bottom-status"
        />
      </div>
    </div>
  </div>
</div>
`;
